mqclient/queue.py

```diff
@@ -327,27 +331,28 @@ async def __aexit__(
         logging.debug(
             f"[MessageAsyncGeneratorContext.__aexit__()] exiting `with-as` block (exc:{exc_type})"
         )
-        if not (self.sub and self.gen):
+        if not (self._sub and self._gen):
             raise RuntimeError(self.RUNTIME_ERROR_CONTEXT_STRING)
 
         reraise_exception = False
 
         # Exception Was Raised
         if exc_type and exc_val:
             if self.msg:
-                await self.queue.nack(self.sub, self.msg)
+                await self.queue._safe_nack(self._sub, self.msg)
             # see how the generator wants to handle the exception
             try:
                 # `athrow` is caught by the generator's try-except around `yield`
-                await self.gen.athrow(exc_type, exc_val, exc_tb)
+                await self._gen.athrow(exc_type, exc_val, exc_tb)
             except exc_type:  # message_generator re-raised Exception
                 reraise_exception = True
         # Good Exit (No Original Exception)
         else:
-            if self.msg:
-                await self.queue.ack(self.sub, self.msg)
+            # ack if there was a message yielded (unless it was already nacked)
+            if self.msg and self.msg._ack_status != Message.AckStatus.NACKED:
```

Comment:
Maybe `_sake_ack` could take a parameter `ignore_nack` that defaults to `False`?

Not sure if it's worth it to make `_safe_ack` more convenient from the calling side, but it may be worth considering.